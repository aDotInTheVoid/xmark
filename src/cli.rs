use clap::Clap;
use eyre::Result;
use std::path::PathBuf;

#[derive(Debug, Default, Clone)]
pub struct Args {
    pub dir: PathBuf,
    pub create: bool,
    pub templates: Option<PathBuf>,
}

impl Args {
    pub fn parse() -> Result<Self> {
        let ArgsInner {
            mut dir,
            create,
            mut templates,
        } = ArgsInner::parse();
        dir = dir.canonicalize()?;
        templates = match templates {
            Some(p) => Some(p.canonicalize()?),
            None => None,
        };
        Ok(Args {
            dir,
            create,
            templates,
        })
    }
}

#[derive(Clap, Debug)]
struct ArgsInner {
    #[clap(short, long, default_value = ".")]
    pub dir: PathBuf,
    #[clap(long)]
    pub create: bool,
    #[clap(short, long)]
    pub templates: Option<PathBuf>,
}
