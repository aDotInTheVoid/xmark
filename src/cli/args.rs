use clap::Clap;
use eyre::Result;
use tracing::instrument;

#[derive(Clap, Debug)]
pub(super) struct Args {
    #[clap(subcommand)]
    subcmd: SubCommand,
}

#[derive(Clap, Debug)]
pub(super) enum SubCommand {
    Build(super::cmd::build::Args), // TODO: init, clean, watch, serve
}

impl Args {
    #[instrument]
    pub fn run(self) -> Result<()> {
        match self.subcmd {
            SubCommand::Build(c) => c.run(),
        }
    }
}
