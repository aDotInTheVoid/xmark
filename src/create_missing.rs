use crate::summary::{Link, Summary};
use eyre::Result;
use std::fs::{self, File};
use std::io::Write;
use std::path::Path;

use log::debug;

pub fn create_missing(src_dir: &Path, summary: &Summary) -> Result<()> {
    let mut items: Vec<_> = summary
        .prefix_chapters
        .iter()
        .chain(summary.suffix_chapters.iter())
        .map(|chapter| Link {
            // This is like 3 strings
            chapter: chapter.clone(),
            ..Default::default()
        })
        //TODO: Don't clone, it can be quite nested
        .chain(summary.numbered_chapters.iter().cloned())
        .collect();

    while !items.is_empty() {
        let link = items.pop().expect("already checked");

        if let Some(ref location) = link.chapter.location {
            let filename = src_dir.join(location);
            if !filename.exists() {
                if let Some(parent) = filename.parent() {
                    if !parent.exists() {
                        fs::create_dir_all(parent)?;
                    }
                }
                debug!("Creating missing file {}", filename.display());

                let mut f = File::create(&filename)?;
                writeln!(f, "# {}", link.chapter.name)?;
            }
        }

        items.extend(link.nested_items);
    }

    Ok(())
}
