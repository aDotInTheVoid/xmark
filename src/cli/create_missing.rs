// SPDX-License-Identifier: GPL-3.0-only
use super::summary::{Link, Summary};
use eyre::Result;
use std::fs::{self, File};
use std::io::Write;
use std::path::Path;

use log::debug;

pub fn create_missing(src_dir: &Path, summary: &Summary) -> Result<()> {
    let mut items: Vec<_> = summary
        .prefix_chapters
        .iter()
        .chain(summary.suffix_chapters.iter())
        .map(|chapter| Link {
            // This is like 3 strings
            chapter: chapter.clone(),
            ..Default::default()
        })
        //TODO: Don't clone, it can be quite nested
        .chain(summary.numbered_chapters.iter().cloned())
        .collect();

    while !items.is_empty() {
        let link = items.pop().expect("already checked");

        if let Some(ref location) = link.chapter.location {
            let filename = src_dir.join(location);
            if !filename.exists() {
                if let Some(parent) = filename.parent() {
                    if !parent.exists() {
                        fs::create_dir_all(parent)?;
                    }
                }
                debug!("Creating missing file {}", filename.display());

                let mut f = File::create(&filename)?;
                writeln!(f, "# {}", link.chapter.name)?;
            }
        }

        items.extend(link.nested_items);
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeSet;
    use std::path::PathBuf;

    use assert_fs::prelude::*;
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::cli::{self, config};
    use crate::html_render;

    #[test]
    fn dummy_e2e() {
        let temp = assert_fs::TempDir::new().unwrap();

        temp.copy_from(
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book"),
            &["xmark.toml", "*/SUMMARY.md"],
        )
        .unwrap();

        let args = cli::Args {
            dir: temp.path().to_owned(),
            create: true,
            ..Default::default()
        };

        let conf = config::load(&args).unwrap();
        let render = html_render::HTMLRender::new(&conf, &args).unwrap();
        render.render().unwrap();

        // BTree so it's in order.
        let paths: BTreeSet<_> = ignore::Walk::new(temp.path())
            .filter_map(|x| x.ok().map(|y| y.into_path()))
            .map(|x| x.into_os_string().into_string().unwrap())
            .filter(|x| !x.contains("_out"))
            .collect();

        assert_yaml_snapshot!(paths, {"[]" => dynamic_redaction(move |val, _| {
            val.as_str().unwrap().replace(temp.path().as_os_str().to_str().unwrap(), "")
        })});
    }
}
