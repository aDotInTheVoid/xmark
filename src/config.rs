use std::fs;
use std::path::PathBuf;

use eyre::{Result, WrapErr};
use serde::{Deserialize, Serialize};

use crate::cli;
use crate::create_missing::create_missing;
use crate::summary::{parse_summary, Chapter, Summary};

/// The Config as represented in the global xmark.toml
#[derive(Clone, Debug, Hash, Serialize, Deserialize)]
pub struct GlobalConfigRepr {
    pub books: Vec<String>,
}

/// The config as usable for the programm
#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
pub struct GlobalConf {
    pub books: Vec<BookConf>,
}

#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
pub struct BookConf {
    pub location: PathBuf,
    pub slug: String,
    pub summary: Summary,
}

pub fn load(args: &cli::Args) -> Result<GlobalConf> {
    let conf = fs::read_to_string(args.dir.clone().join("xmark.toml"))
        .with_context(|| "Couldn't find xmark.toml")?;
    let conf: GlobalConfigRepr = toml::from_str(&conf)?;

    hydrate(conf, &args)
}

// Convert the disk format to a usable form
pub fn hydrate(gcr: GlobalConfigRepr, args: &cli::Args) -> Result<GlobalConf> {
    Ok(GlobalConf {
        books: gcr
            .books
            .iter()
            .map(|name| {
                let location = args.dir.join(name);
                let summary_location = location.join("SUMMARY.md");
                let mut summary = parse_summary(
                    &fs::read_to_string(&summary_location)
                        .wrap_err_with(|| format!("Couldn't open {:?}", summary_location))?,
                )?;

                if args.create {
                    create_missing(&location, &summary)?;
                }

                let fix_chap_loc = |chap: &mut Chapter| {
                    if let Some(loc) = chap.location.as_deref() {
                        chap.location = Some(location.join(loc));
                    }
                };

                summary.prefix_chapters.iter_mut().for_each(fix_chap_loc);

                summary.suffix_chapters.iter_mut().for_each(fix_chap_loc);

                summary.numbered_chapters.iter_mut().for_each(|chap| {
                    chap.map_mut(fix_chap_loc);
                });

                Ok(BookConf {
                    location,
                    summary,
                    slug: name.to_owned(),
                })
            })
            .collect::<Result<_>>()?,
    })
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;

    #[test]
    fn de_gloal_config() {
        let inp = "books = [
            'book-1',
            'book-2',
            'book-3',
            ]
";
        let conf: GlobalConfigRepr = toml::from_str(inp).unwrap();
        assert_eq!(conf.books, vec!["book-1", "book-2", "book-3"]);
    }

    #[test]
    fn hydrate_basic() {
        let args = cli::Args {
            dir: "/home/etc/bax".into(),
            create: false,
        };
        let gcr = GlobalConfigRepr { books: vec![] };
        let gc = GlobalConf { books: vec![] };
        assert_eq!(hydrate(gcr, &args).unwrap(), gc);
    }

    #[test]
    fn hydrate_dummy() {
        // I don't know why I need the anotation
        let redacter = |mut val, _: insta::internals::ContentPath| {
            // TODO: Dont do insta crimes.
            // This is in #[doc(Hidden)] internals.
            while let insta::internals::Content::Some(some) = val {
                val = *some;
            }

            if let insta::internals::Content::String(s) = val {
                s.replace(dbg!(env!("CARGO_MANIFEST_DIR")), "BASEDIR")
            } else {
                "".into()
            }
        };

        let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book");
        let args = cli::Args { dir, create: false };
        let conf = load(&args).unwrap();
        assert_yaml_snapshot!(conf, {
            ".location" => insta::dynamic_redaction(redacter),
            ".**.location" => insta::dynamic_redaction(redacter),
        });
    }
}
