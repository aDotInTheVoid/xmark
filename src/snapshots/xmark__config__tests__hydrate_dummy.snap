---
source: src/config.rs
expression: conf
---
books:
  - location: BASEDIR/dummy-book/book-1
    summary:
      title: The Rust Programming Language
      prefix_chapters:
        - name: The Rust Programming Language
          location: BASEDIR/dummy-book/book-1/title-page.md
        - name: Foreword
          location: BASEDIR/dummy-book/book-1/foreword.md
        - name: Introduction
          location: BASEDIR/dummy-book/book-1/ch00-00-introduction.md
      numbered_chapters:
        - chapter:
            name: Getting Started
            location: BASEDIR/dummy-book/book-1/ch01-00-getting-started.md
          nested_items:
            - chapter:
                name: Installation
                location: BASEDIR/dummy-book/book-1/ch01-01-installation.md
              nested_items: []
              section_number:
                - 1
                - 1
            - chapter:
                name: "Hello, World!"
                location: BASEDIR/dummy-book/book-1/ch01-02-hello-world.md
              nested_items: []
              section_number:
                - 1
                - 2
            - chapter:
                name: "Hello, Cargo!"
                location: BASEDIR/dummy-book/book-1/ch01-03-hello-cargo.md
              nested_items: []
              section_number:
                - 1
                - 3
          section_number:
            - 1
        - chapter:
            name: Programming a Guessing Game
            location: BASEDIR/dummy-book/book-1/ch02-00-guessing-game-tutorial.md
          nested_items: []
          section_number:
            - 2
        - chapter:
            name: Common Programming Concepts
            location: BASEDIR/dummy-book/book-1/ch03-00-common-programming-concepts.md
          nested_items:
            - chapter:
                name: Variables and Mutability
                location: BASEDIR/dummy-book/book-1/ch03-01-variables-and-mutability.md
              nested_items: []
              section_number:
                - 3
                - 1
            - chapter:
                name: Data Types
                location: BASEDIR/dummy-book/book-1/ch03-02-data-types.md
              nested_items: []
              section_number:
                - 3
                - 2
            - chapter:
                name: Functions
                location: BASEDIR/dummy-book/book-1/ch03-03-how-functions-work.md
              nested_items: []
              section_number:
                - 3
                - 3
            - chapter:
                name: Comments
                location: BASEDIR/dummy-book/book-1/ch03-04-comments.md
              nested_items: []
              section_number:
                - 3
                - 4
            - chapter:
                name: Control Flow
                location: BASEDIR/dummy-book/book-1/ch03-05-control-flow.md
              nested_items: []
              section_number:
                - 3
                - 5
          section_number:
            - 3
        - chapter:
            name: Understanding Ownership
            location: BASEDIR/dummy-book/book-1/ch04-00-understanding-ownership.md
          nested_items:
            - chapter:
                name: What is Ownership?
                location: BASEDIR/dummy-book/book-1/ch04-01-what-is-ownership.md
              nested_items: []
              section_number:
                - 4
                - 1
            - chapter:
                name: References and Borrowing
                location: BASEDIR/dummy-book/book-1/ch04-02-references-and-borrowing.md
              nested_items: []
              section_number:
                - 4
                - 2
            - chapter:
                name: The Slice Type
                location: BASEDIR/dummy-book/book-1/ch04-03-slices.md
              nested_items: []
              section_number:
                - 4
                - 3
          section_number:
            - 4
        - chapter:
            name: Using Structs to Structure Related Data
            location: BASEDIR/dummy-book/book-1/ch05-00-structs.md
          nested_items:
            - chapter:
                name: Defining and Instantiating Structs
                location: BASEDIR/dummy-book/book-1/ch05-01-defining-structs.md
              nested_items: []
              section_number:
                - 5
                - 1
            - chapter:
                name: An Example Program Using Structs
                location: BASEDIR/dummy-book/book-1/ch05-02-example-structs.md
              nested_items: []
              section_number:
                - 5
                - 2
            - chapter:
                name: Method Syntax
                location: BASEDIR/dummy-book/book-1/ch05-03-method-syntax.md
              nested_items: []
              section_number:
                - 5
                - 3
          section_number:
            - 5
        - chapter:
            name: Enums and Pattern Matching
            location: BASEDIR/dummy-book/book-1/ch06-00-enums.md
          nested_items:
            - chapter:
                name: Defining an Enum
                location: BASEDIR/dummy-book/book-1/ch06-01-defining-an-enum.md
              nested_items: []
              section_number:
                - 6
                - 1
            - chapter:
                name: The match Control Flow Operator
                location: BASEDIR/dummy-book/book-1/ch06-02-match.md
              nested_items: []
              section_number:
                - 6
                - 2
            - chapter:
                name: Concise Control Flow with if let
                location: BASEDIR/dummy-book/book-1/ch06-03-if-let.md
              nested_items: []
              section_number:
                - 6
                - 3
          section_number:
            - 6
        - chapter:
            name: "Managing Growing Projects with Packages, Crates, and Modules"
            location: BASEDIR/dummy-book/book-1/ch07-00-managing-growing-projects-with-packages-crates-and-modules.md
          nested_items:
            - chapter:
                name: Packages and Crates
                location: BASEDIR/dummy-book/book-1/ch07-01-packages-and-crates.md
              nested_items: []
              section_number:
                - 7
                - 1
            - chapter:
                name: Defining Modules to Control Scope and Privacy
                location: BASEDIR/dummy-book/book-1/ch07-02-defining-modules-to-control-scope-and-privacy.md
              nested_items: []
              section_number:
                - 7
                - 2
            - chapter:
                name: Paths for Referring to an Item in the Module Tree
                location: BASEDIR/dummy-book/book-1/ch07-03-paths-for-referring-to-an-item-in-the-module-tree.md
              nested_items: []
              section_number:
                - 7
                - 3
            - chapter:
                name: Bringing Paths Into Scope with the use Keyword
                location: BASEDIR/dummy-book/book-1/ch07-04-bringing-paths-into-scope-with-the-use-keyword.md
              nested_items: []
              section_number:
                - 7
                - 4
            - chapter:
                name: Separating Modules into Different Files
                location: BASEDIR/dummy-book/book-1/ch07-05-separating-modules-into-different-files.md
              nested_items: []
              section_number:
                - 7
                - 5
          section_number:
            - 7
        - chapter:
            name: Common Collections
            location: BASEDIR/dummy-book/book-1/ch08-00-common-collections.md
          nested_items:
            - chapter:
                name: Storing Lists of Values with Vectors
                location: BASEDIR/dummy-book/book-1/ch08-01-vectors.md
              nested_items: []
              section_number:
                - 8
                - 1
            - chapter:
                name: Storing UTF-8 Encoded Text with Strings
                location: BASEDIR/dummy-book/book-1/ch08-02-strings.md
              nested_items: []
              section_number:
                - 8
                - 2
            - chapter:
                name: Storing Keys with Associated Values in Hash Maps
                location: BASEDIR/dummy-book/book-1/ch08-03-hash-maps.md
              nested_items: []
              section_number:
                - 8
                - 3
          section_number:
            - 8
        - chapter:
            name: Error Handling
            location: BASEDIR/dummy-book/book-1/ch09-00-error-handling.md
          nested_items:
            - chapter:
                name: Unrecoverable Errors with panic!
                location: BASEDIR/dummy-book/book-1/ch09-01-unrecoverable-errors-with-panic.md
              nested_items: []
              section_number:
                - 9
                - 1
            - chapter:
                name: Recoverable Errors with Result
                location: BASEDIR/dummy-book/book-1/ch09-02-recoverable-errors-with-result.md
              nested_items: []
              section_number:
                - 9
                - 2
            - chapter:
                name: To panic! or Not To panic!
                location: BASEDIR/dummy-book/book-1/ch09-03-to-panic-or-not-to-panic.md
              nested_items: []
              section_number:
                - 9
                - 3
          section_number:
            - 9
        - chapter:
            name: "Generic Types, Traits, and Lifetimes"
            location: BASEDIR/dummy-book/book-1/ch10-00-generics.md
          nested_items:
            - chapter:
                name: Generic Data Types
                location: BASEDIR/dummy-book/book-1/ch10-01-syntax.md
              nested_items: []
              section_number:
                - 10
                - 1
            - chapter:
                name: "Traits: Defining Shared Behavior"
                location: BASEDIR/dummy-book/book-1/ch10-02-traits.md
              nested_items: []
              section_number:
                - 10
                - 2
            - chapter:
                name: Validating References with Lifetimes
                location: BASEDIR/dummy-book/book-1/ch10-03-lifetime-syntax.md
              nested_items: []
              section_number:
                - 10
                - 3
          section_number:
            - 10
        - chapter:
            name: Writing Automated Tests
            location: BASEDIR/dummy-book/book-1/ch11-00-testing.md
          nested_items:
            - chapter:
                name: How to Write Tests
                location: BASEDIR/dummy-book/book-1/ch11-01-writing-tests.md
              nested_items: []
              section_number:
                - 11
                - 1
            - chapter:
                name: Controlling How Tests Are Run
                location: BASEDIR/dummy-book/book-1/ch11-02-running-tests.md
              nested_items: []
              section_number:
                - 11
                - 2
            - chapter:
                name: Test Organization
                location: BASEDIR/dummy-book/book-1/ch11-03-test-organization.md
              nested_items: []
              section_number:
                - 11
                - 3
          section_number:
            - 11
        - chapter:
            name: "An I/O Project: Building a Command Line Program"
            location: BASEDIR/dummy-book/book-1/ch12-00-an-io-project.md
          nested_items:
            - chapter:
                name: Accepting Command Line Arguments
                location: BASEDIR/dummy-book/book-1/ch12-01-accepting-command-line-arguments.md
              nested_items: []
              section_number:
                - 12
                - 1
            - chapter:
                name: Reading a File
                location: BASEDIR/dummy-book/book-1/ch12-02-reading-a-file.md
              nested_items: []
              section_number:
                - 12
                - 2
            - chapter:
                name: Refactoring to Improve Modularity and Error Handling
                location: BASEDIR/dummy-book/book-1/ch12-03-improving-error-handling-and-modularity.md
              nested_items: []
              section_number:
                - 12
                - 3
            - chapter:
                name: Developing the Library’s Functionality with Test Driven Development
                location: BASEDIR/dummy-book/book-1/ch12-04-testing-the-librarys-functionality.md
              nested_items: []
              section_number:
                - 12
                - 4
            - chapter:
                name: Working with Environment Variables
                location: BASEDIR/dummy-book/book-1/ch12-05-working-with-environment-variables.md
              nested_items: []
              section_number:
                - 12
                - 5
            - chapter:
                name: Writing Error Messages to Standard Error Instead of Standard Output
                location: BASEDIR/dummy-book/book-1/ch12-06-writing-to-stderr-instead-of-stdout.md
              nested_items: []
              section_number:
                - 12
                - 6
          section_number:
            - 12
        - chapter:
            name: "Functional Language Features: Iterators and Closures"
            location: BASEDIR/dummy-book/book-1/ch13-00-functional-features.md
          nested_items:
            - chapter:
                name: "Closures: Anonymous Functions that Can Capture Their Environment"
                location: BASEDIR/dummy-book/book-1/ch13-01-closures.md
              nested_items: []
              section_number:
                - 13
                - 1
            - chapter:
                name: Processing a Series of Items with Iterators
                location: BASEDIR/dummy-book/book-1/ch13-02-iterators.md
              nested_items: []
              section_number:
                - 13
                - 2
            - chapter:
                name: Improving Our I/O Project
                location: BASEDIR/dummy-book/book-1/ch13-03-improving-our-io-project.md
              nested_items: []
              section_number:
                - 13
                - 3
            - chapter:
                name: "Comparing Performance: Loops vs. Iterators"
                location: BASEDIR/dummy-book/book-1/ch13-04-performance.md
              nested_items: []
              section_number:
                - 13
                - 4
          section_number:
            - 13
        - chapter:
            name: More about Cargo and Crates.io
            location: BASEDIR/dummy-book/book-1/ch14-00-more-about-cargo.md
          nested_items:
            - chapter:
                name: Customizing Builds with Release Profiles
                location: BASEDIR/dummy-book/book-1/ch14-01-release-profiles.md
              nested_items: []
              section_number:
                - 14
                - 1
            - chapter:
                name: Publishing a Crate to Crates.io
                location: BASEDIR/dummy-book/book-1/ch14-02-publishing-to-crates-io.md
              nested_items: []
              section_number:
                - 14
                - 2
            - chapter:
                name: Cargo Workspaces
                location: BASEDIR/dummy-book/book-1/ch14-03-cargo-workspaces.md
              nested_items: []
              section_number:
                - 14
                - 3
            - chapter:
                name: Installing Binaries from Crates.io with cargo install
                location: BASEDIR/dummy-book/book-1/ch14-04-installing-binaries.md
              nested_items: []
              section_number:
                - 14
                - 4
            - chapter:
                name: Extending Cargo with Custom Commands
                location: BASEDIR/dummy-book/book-1/ch14-05-extending-cargo.md
              nested_items: []
              section_number:
                - 14
                - 5
          section_number:
            - 14
        - chapter:
            name: Smart Pointers
            location: BASEDIR/dummy-book/book-1/ch15-00-smart-pointers.md
          nested_items:
            - chapter:
                name: Using Box<T> to Point to Data on the Heap
                location: BASEDIR/dummy-book/book-1/ch15-01-box.md
              nested_items: []
              section_number:
                - 15
                - 1
            - chapter:
                name: Treating Smart Pointers Like Regular References with the Deref Trait
                location: BASEDIR/dummy-book/book-1/ch15-02-deref.md
              nested_items: []
              section_number:
                - 15
                - 2
            - chapter:
                name: Running Code on Cleanup with the Drop Trait
                location: BASEDIR/dummy-book/book-1/ch15-03-drop.md
              nested_items: []
              section_number:
                - 15
                - 3
            - chapter:
                name: "Rc<T>, the Reference Counted Smart Pointer"
                location: BASEDIR/dummy-book/book-1/ch15-04-rc.md
              nested_items: []
              section_number:
                - 15
                - 4
            - chapter:
                name: RefCell<T> and the Interior Mutability Pattern
                location: BASEDIR/dummy-book/book-1/ch15-05-interior-mutability.md
              nested_items: []
              section_number:
                - 15
                - 5
            - chapter:
                name: Reference Cycles Can Leak Memory
                location: BASEDIR/dummy-book/book-1/ch15-06-reference-cycles.md
              nested_items: []
              section_number:
                - 15
                - 6
          section_number:
            - 15
        - chapter:
            name: Fearless Concurrency
            location: BASEDIR/dummy-book/book-1/ch16-00-concurrency.md
          nested_items:
            - chapter:
                name: Using Threads to Run Code Simultaneously
                location: BASEDIR/dummy-book/book-1/ch16-01-threads.md
              nested_items: []
              section_number:
                - 16
                - 1
            - chapter:
                name: Using Message Passing to Transfer Data Between Threads
                location: BASEDIR/dummy-book/book-1/ch16-02-message-passing.md
              nested_items: []
              section_number:
                - 16
                - 2
            - chapter:
                name: Shared-State Concurrency
                location: BASEDIR/dummy-book/book-1/ch16-03-shared-state.md
              nested_items: []
              section_number:
                - 16
                - 3
            - chapter:
                name: Extensible Concurrency with the Sync and Send Traits
                location: BASEDIR/dummy-book/book-1/ch16-04-extensible-concurrency-sync-and-send.md
              nested_items: []
              section_number:
                - 16
                - 4
          section_number:
            - 16
        - chapter:
            name: Object Oriented Programming Features of Rust
            location: BASEDIR/dummy-book/book-1/ch17-00-oop.md
          nested_items:
            - chapter:
                name: Characteristics of Object-Oriented Languages
                location: BASEDIR/dummy-book/book-1/ch17-01-what-is-oo.md
              nested_items: []
              section_number:
                - 17
                - 1
            - chapter:
                name: Using Trait Objects That Allow for Values of Different Types
                location: BASEDIR/dummy-book/book-1/ch17-02-trait-objects.md
              nested_items: []
              section_number:
                - 17
                - 2
            - chapter:
                name: Implementing an Object-Oriented Design Pattern
                location: BASEDIR/dummy-book/book-1/ch17-03-oo-design-patterns.md
              nested_items: []
              section_number:
                - 17
                - 3
          section_number:
            - 17
        - chapter:
            name: Patterns and Matching
            location: BASEDIR/dummy-book/book-1/ch18-00-patterns.md
          nested_items:
            - chapter:
                name: All the Places Patterns Can Be Used
                location: BASEDIR/dummy-book/book-1/ch18-01-all-the-places-for-patterns.md
              nested_items: []
              section_number:
                - 18
                - 1
            - chapter:
                name: "Refutability: Whether a Pattern Might Fail to Match"
                location: BASEDIR/dummy-book/book-1/ch18-02-refutability.md
              nested_items: []
              section_number:
                - 18
                - 2
            - chapter:
                name: Pattern Syntax
                location: BASEDIR/dummy-book/book-1/ch18-03-pattern-syntax.md
              nested_items: []
              section_number:
                - 18
                - 3
          section_number:
            - 18
        - chapter:
            name: Advanced Features
            location: BASEDIR/dummy-book/book-1/ch19-00-advanced-features.md
          nested_items:
            - chapter:
                name: Unsafe Rust
                location: BASEDIR/dummy-book/book-1/ch19-01-unsafe-rust.md
              nested_items: []
              section_number:
                - 19
                - 1
            - chapter:
                name: Advanced Traits
                location: BASEDIR/dummy-book/book-1/ch19-03-advanced-traits.md
              nested_items: []
              section_number:
                - 19
                - 2
            - chapter:
                name: Advanced Types
                location: BASEDIR/dummy-book/book-1/ch19-04-advanced-types.md
              nested_items: []
              section_number:
                - 19
                - 3
            - chapter:
                name: Advanced Functions and Closures
                location: BASEDIR/dummy-book/book-1/ch19-05-advanced-functions-and-closures.md
              nested_items: []
              section_number:
                - 19
                - 4
            - chapter:
                name: Macros
                location: BASEDIR/dummy-book/book-1/ch19-06-macros.md
              nested_items: []
              section_number:
                - 19
                - 5
          section_number:
            - 19
        - chapter:
            name: "Final Project: Building a Multithreaded Web Server"
            location: BASEDIR/dummy-book/book-1/ch20-00-final-project-a-web-server.md
          nested_items:
            - chapter:
                name: Building a Single-Threaded Web Server
                location: BASEDIR/dummy-book/book-1/ch20-01-single-threaded.md
              nested_items: []
              section_number:
                - 20
                - 1
            - chapter:
                name: Turning Our Single-Threaded Server into a Multithreaded Server
                location: BASEDIR/dummy-book/book-1/ch20-02-multithreaded.md
              nested_items: []
              section_number:
                - 20
                - 2
            - chapter:
                name: Graceful Shutdown and Cleanup
                location: BASEDIR/dummy-book/book-1/ch20-03-graceful-shutdown-and-cleanup.md
              nested_items: []
              section_number:
                - 20
                - 3
          section_number:
            - 20
        - chapter:
            name: Appendix
            location: BASEDIR/dummy-book/book-1/appendix-00.md
          nested_items:
            - chapter:
                name: A - Keywords
                location: BASEDIR/dummy-book/book-1/appendix-01-keywords.md
              nested_items: []
              section_number:
                - 21
                - 1
            - chapter:
                name: B - Operators and Symbols
                location: BASEDIR/dummy-book/book-1/appendix-02-operators.md
              nested_items: []
              section_number:
                - 21
                - 2
            - chapter:
                name: C - Derivable Traits
                location: BASEDIR/dummy-book/book-1/appendix-03-derivable-traits.md
              nested_items: []
              section_number:
                - 21
                - 3
            - chapter:
                name: D - Useful Development Tools
                location: BASEDIR/dummy-book/book-1/appendix-04-useful-development-tools.md
              nested_items: []
              section_number:
                - 21
                - 4
            - chapter:
                name: E - Editions
                location: BASEDIR/dummy-book/book-1/appendix-05-editions.md
              nested_items: []
              section_number:
                - 21
                - 5
            - chapter:
                name: F - Translations of the Book
                location: BASEDIR/dummy-book/book-1/appendix-06-translation.md
              nested_items: []
              section_number:
                - 21
                - 6
            - chapter:
                name: G - How Rust is Made and “Nightly Rust”
                location: BASEDIR/dummy-book/book-1/appendix-07-nightly-rust.md
              nested_items: []
              section_number:
                - 21
                - 7
          section_number:
            - 21
      suffix_chapters: []
  - location: BASEDIR/dummy-book/book-2
    summary:
      title: Summary
      prefix_chapters: []
      numbered_chapters:
        - chapter:
            name: mdBook
            location: BASEDIR/dummy-book/book-2/README.md
          nested_items: []
          section_number:
            - 1
        - chapter:
            name: Command Line Tool
            location: BASEDIR/dummy-book/book-2/cli/README.md
          nested_items:
            - chapter:
                name: init
                location: BASEDIR/dummy-book/book-2/cli/init.md
              nested_items: []
              section_number:
                - 2
                - 1
            - chapter:
                name: build
                location: BASEDIR/dummy-book/book-2/cli/build.md
              nested_items: []
              section_number:
                - 2
                - 2
            - chapter:
                name: watch
                location: BASEDIR/dummy-book/book-2/cli/watch.md
              nested_items: []
              section_number:
                - 2
                - 3
            - chapter:
                name: serve
                location: BASEDIR/dummy-book/book-2/cli/serve.md
              nested_items: []
              section_number:
                - 2
                - 4
            - chapter:
                name: test
                location: BASEDIR/dummy-book/book-2/cli/test.md
              nested_items: []
              section_number:
                - 2
                - 5
            - chapter:
                name: clean
                location: BASEDIR/dummy-book/book-2/cli/clean.md
              nested_items: []
              section_number:
                - 2
                - 6
          section_number:
            - 2
        - chapter:
            name: Format
            location: BASEDIR/dummy-book/book-2/format/README.md
          nested_items:
            - chapter:
                name: SUMMARY.md
                location: BASEDIR/dummy-book/book-2/format/summary.md
              nested_items:
                - chapter:
                    name: Draft chapter
                    location: ~
                  nested_items: []
                  section_number:
                    - 3
                    - 1
                    - 1
              section_number:
                - 3
                - 1
            - chapter:
                name: Configuration
                location: BASEDIR/dummy-book/book-2/format/config.md
              nested_items: []
              section_number:
                - 3
                - 2
            - chapter:
                name: Theme
                location: BASEDIR/dummy-book/book-2/format/theme/README.md
              nested_items:
                - chapter:
                    name: index.hbs
                    location: BASEDIR/dummy-book/book-2/format/theme/index-hbs.md
                  nested_items: []
                  section_number:
                    - 3
                    - 3
                    - 1
                - chapter:
                    name: Syntax highlighting
                    location: BASEDIR/dummy-book/book-2/format/theme/syntax-highlighting.md
                  nested_items: []
                  section_number:
                    - 3
                    - 3
                    - 2
                - chapter:
                    name: Editor
                    location: BASEDIR/dummy-book/book-2/format/theme/editor.md
                  nested_items: []
                  section_number:
                    - 3
                    - 3
                    - 3
              section_number:
                - 3
                - 3
            - chapter:
                name: MathJax Support
                location: BASEDIR/dummy-book/book-2/format/mathjax.md
              nested_items: []
              section_number:
                - 3
                - 4
            - chapter:
                name: mdBook-specific features
                location: BASEDIR/dummy-book/book-2/format/mdbook.md
              nested_items: []
              section_number:
                - 3
                - 5
          section_number:
            - 3
        - chapter:
            name: Continuous Integration
            location: BASEDIR/dummy-book/book-2/continuous-integration.md
          nested_items: []
          section_number:
            - 4
        - chapter:
            name: For Developers
            location: BASEDIR/dummy-book/book-2/for_developers/README.md
          nested_items:
            - chapter:
                name: Preprocessors
                location: BASEDIR/dummy-book/book-2/for_developers/preprocessors.md
              nested_items: []
              section_number:
                - 5
                - 1
            - chapter:
                name: Alternative Backends
                location: BASEDIR/dummy-book/book-2/for_developers/backends.md
              nested_items: []
              section_number:
                - 5
                - 2
          section_number:
            - 5
      suffix_chapters:
        - name: Contributors
          location: BASEDIR/dummy-book/book-2/misc/contributors.md
  - location: BASEDIR/dummy-book/book-3
    summary:
      title: "3"
      prefix_chapters:
        - name: pre 1
          location: BASEDIR/dummy-book/book-3/./pre1.md
        - name: pre 2
          location: BASEDIR/dummy-book/book-3/./pre2.md
      numbered_chapters:
        - chapter:
            name: c 1
            location: BASEDIR/dummy-book/book-3/./c1.md
          nested_items: []
          section_number:
            - 1
        - chapter:
            name: c 2
            location: BASEDIR/dummy-book/book-3/./c2.md
          nested_items: []
          section_number:
            - 2
        - chapter:
            name: c 3
            location: BASEDIR/dummy-book/book-3/./c3.md
          nested_items:
            - chapter:
                name: c 3 1
                location: BASEDIR/dummy-book/book-3/./c31.md
              nested_items: []
              section_number:
                - 3
                - 1
            - chapter:
                name: c 3 2
                location: BASEDIR/dummy-book/book-3/./c32.md
              nested_items: []
              section_number:
                - 3
                - 2
            - chapter:
                name: c 3 3
                location: BASEDIR/dummy-book/book-3/./c33.md
              nested_items: []
              section_number:
                - 3
                - 3
          section_number:
            - 3
        - chapter:
            name: post 1
            location: BASEDIR/dummy-book/book-3/./post1.md
          nested_items: []
          section_number:
            - 4
      suffix_chapters: []
html:
  site-url: ~
